keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+  discusses the problem of blocking PDBs and offers several workarounds. Similar to others also
+  offers a force deletion, but also a less destructive method of scaling up the application.
+  However, this also interferes with application deployment and has to be supported by the
+  application.
+
+### Cluster Autoscaler
+
+Accepts a `drain-priority-config` option, which is similar to Graceful Node Shutdown in that it
+gives each priority a shutdown grace period. Also has a `max-graceful-termination-sec` option for
+pod termination and a `max-pod-eviction-time` option after which the eviction is forfeited.
+
+Each pod is first analyzed to see if it is drainable. Part of the logic is similar to kubectl and
+its drain filters (see [Cluster Autoscaler rules](https://github.com/kubernetes/autoscaler/blob/554366f979b11aeb82df335a793e4d7a1acfadb4/cluster-autoscaler/simulator/drainability/rules/rules.go#L50-L77)):
+- Mirror pods are skipped.
+- Terminating pods are skipped.
+- Pods and ReplicaSets/ReplicationControllers without owning controllers are blocking by default
+  (the check can be modified with the `skip-nodes-with-custom-controller-pods` option).
+- System pods (in the `kube-system` namespace) without a matching PDB are blocking by default
+  (the check can be modified with the `skip-nodes-with-system-pods` option).
+- Pods with `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` annotation are blocking.
+- Pods with local storage are blocking unless they have a
+  `cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes` annotation
+  (the check can be modified with the `skip-nodes-with-local-storage` option, e.g., this check
+  is skipped on [AKS](https://learn.microsoft.com/en-us/azure/aks/cluster-autoscaler?tabs=azure-cli#cluster-autoscaler-profile-settings)).
+- Pods with PDBs that do not have `disruptionsAllowed` are blocking.
+
+This can be enhanced with other rules and overrides.
+
+It uses this logic to first check if all pods can be removed from a node. If not, it will report
+those nodes. Then it will group all pods by a priority and evict them gradually from the lowest
+to highest priority. This may include DaemonSet pods.
+
+### kubelet
+
+Graceful Node Shutdown is a part of the current solution for node maintenance. Unfortunately, it is
+not possible to rely solely on this feature as a go-to solution for graceful node and workload
+termination.
+
+- The Graceful Node Shutdown feature is not application aware and may prematurely disrupt workloads 
+  and lead to data loss.
+- The kubelet controls the shutdown process using Dbus and systemd, and can delay (but not entirely
+  block) it using the systemd inhibitor. However, if Dbus or the kubelet is restarted during the
+  node shutdown, the shutdown might not be registered again, and pods might be terminated
+  ungracefully. Also, new workloads can get scheduled on the node while the node is shutting down.
```

Comment:
This seems more like a mismanaged node maintenance, that no solution can fix for you, other than having a strict runbook you follow when performing one.

Reply:
Adding an API to mark that the node is covered by a maintenance period should help here; it'd be a declarative way for not just tainting nodes but also allowing workload controllers to surge before the taint actually triggers eviction. The API I'd imagined provides more context than a taint on its own can.

Either the scheduler pays attention to node maintenance or learns to take account of a derived taint, but either way there should be a way to either prefer or enforce that new Pods don't get bound onto a node due to shut down imminently.

Reply:
I don't understand why the indirection on a third API object it is better, can not we add this state to the node object itself? we already have taints and conditions that indicate the node is not ready or it can not receive pods

Reply:
Drawbacks of adding to Node:
- one of the ideas was to define a _selector_ matching multiple nodes under maintenance. If you put it into the `.spec` of the Node, then something has to set / manage that spec
- you can have two or more concurrent reasons to mark a node as being liable to disruption
  1. example: the node autoscaler for that cluster no longer wants the node to exist
  2. also, that zone is being emptied in response to a provider-level maintenance notification
  3. the node problem detector has also triggered a fault report on this node
  
  using Server-Side Apply and multiple managers on fields is feasible, but puts the onus on each client to play well with others.
- we can revise the implementation at alpha and rely on clients to only integrate with implementations they know they support, using a match on the API version(s) served for NodeMaintenance. If we add the API to Node, clients don't have that option. The Node API will be available at v1 indefinitely and the meaning of a particular field is stable the moment we add it, pretty much.
- with a taint, we can't record a reason. If we add an official equivalent of NodeMaintenance from https://github.com/medik8s/node-maintenance-operator, we can.
  - I'm inclined to offer optional start and end times (UTC) for maintenance, because that matches real-world needs.

Reply:
> one of the ideas was to define a selector matching multiple nodes under maintenance. If you put it into the .spec of the Node, then something has to set / manage that spec

you are defining a state on the node, spec seems the right place

> you can have two or more concurrent reasons to mark a node as being liable to disruption

exactly there is only one Node and one state, so is the object that provides the synchronization, I may miss the use case of playing well together, the lifecycle of the node should respect a state machine

> we can revise the implementation

Node should only have the necessary primitives ... and fix the bugs and limitations that exist today, that need some work  as @soltysh mentioned in other comments .Building compensation layers is easier in the short term but ends adding more technical debt,

 if the lifecycle of the nodes and the primitives implemented on the Node exist, we don't need to absorb the complexity of an orchestrator in tree with all the actuators and evacuators and .... 

> with a taint, we can't record a reason. If we add an official equivalent of NodeMaintenance from https://github.com/medik8s/node-maintenance-operator, we can.

> I'm inclined to offer optional start and end times (UTC) for maintenance, because that matches real-world needs.

This start to move to the area of orchestration and day 2 operations, does not feel something for core, and more for ClusterAPI or another projects on that area

Reply:
It is hard to achieve synchronization of a state through a single object, if you have multiple actors requesting the maintenance and then cancelling it at different times. We are talking about admins sending patch requests, not a handful of well-behaved components. 

We have the same problem today with cordon/uncordon (`.spec.Unschedulable`) today. See https://github.com/kubernetes/enhancements/pull/4213#discussion_r1537798784. This KEP also has to handle this field in a special manner. If it becomes a problem, we can detect NM in the scheduler.


We can move some part of the state/status to the node as mentioned in https://github.com/kubernetes/enhancements/pull/4213#discussion_r1631539092. But it is still useful to have the NodeMaintenance API to decouple the ownership of the maintenance and also to decouple the functional aspects.

The evacuation API is also designed with the concurrency in mind. The API is not only useful for the NodeMaintenance, but has much broader use cases (comparable/larger than eviction). It provides a way for any actor to observe the eviction/evacuation and thus it is possible to gracefully remove any type of pod (normal pod, daemonset, static, pod under a PDB, pod not under a PDB, pod of a workload that does not support PDBs, pod in any state). Applications would have full control over the lifecycle of their own pods. This is not possible today, and people are forced to write a lot of workarounds to get closer to the desired behavior.

Later, it should be possible to extend the Evacuation API if needed, to gracefully remove any kind of a resource.

Reply:
> you are defining a state on the node, spec seems the right place

I'm expecting to define a desired state across a _set_ of nodes (potentially 4999 of them: our soft limit on node scaling, minus one). If I can define that in one API call, it's easier than making 4999 calls and the client gets to return a single success failure.